lib/spack/spack/stage.py

```diff
@@ -725,6 +739,154 @@ def _add_to_root_stage(self):
+        self._hash = hash
+        self._remote_build_dir = None
+        self._remote_root = Path(spack.config.get("config:cmake_ext_build"))
+
+    @property
+    def _root_stage_context(self):
+        return self._path / ("spack-build-%s" % self._hash)
+
+    @property
+    def build_directory(self):
+        return self._remote_build_dir
+
+    def _establish_context_link(self):
+        symlink(str(self._remote_build_dir), str(self._root_stage_context))
+
+    def _remove_context_link(self):
+        self._root_stage_context.unlink()
+
+    def _setup_remote_build_stage(self):
+        # try to create root if it doesn't exist
+        self._remote_root.mkdir(parents=True, exist_ok=True)
+        sub_dir = self._compute_next_open_subdir()
+        full_subdir = self._remote_root / sub_dir
+        full_subdir.mkdir()
+        return full_subdir
+
+    def _compute_next_open_subdir(self, last=None):
+        def inc(c):
+            curr = ord(c[-1])
+            over = curr // 122
+            if over:
+                return (inc(c[:-1]) if len(c[:-1]) else "a") + "a"
+            return c[:-1] + chr(curr + 1)
+
+        sort_key = lambda x: (len(x.name), x.name)
+        current_ext_stages = list(self._remote_root.iterdir())
+        if not current_ext_stages:
+            # no currently extant external stages, start enumerating with 'a'
+            return "a"
+        last = sorted(current_ext_stages, key=sort_key)[-1]
+        return inc(last.name)
+
+    def _teardown_remote_build_dir(self):
+        """Destroy external build tree if not keep-stage
+        Otherwise this is kept as usual"""
+
+        def teardown(pth: Path):
+            for sub_item in pth.iterdir():
+                if sub_item.is_dir():
+                    teardown(sub_item)
+                else:
+                    sub_item.unlink()
+            pth.rmdir()
+
+        if self._remote_build_dir.exists():
+            teardown(self._remote_build_dir)
+        self._remote_build_dir = None
+
+    def _destroy_remote(self):
+        # copy back to stage may fail in event of error, make sure we clean up the
+        # associated external build dir in that event unless we're keeping the
+        # parent stage on cleanup
+        # If remote stage is not set, we never created one, package is already
+        # installed and we should do nothing here
+        if self._remote_build_dir:
+            try:
+                self._remove_context_link()
+            finally:
+                self._teardown_remote_build_dir()
+
+    def contains(self, file_path):
+        """Returns true if file_path is in root stage
+        or the remote build dir"""
+        # Check if file is in root stage
+        cont = super(CMakeBuildStage, self).contains(file_path)
+        # if not, check if it's in remote build dir
+        if not cont:
+            cont = str(self._remote_build_dir) in file_path
+        return cont
+
+    def create(self):
+        if not self.created:
```

Comment:
I'll leave the `try/except` as if there is an exception in the remote build dir creation/link establishment, if the remote dir is created, it will not be cleaned up. The rest of the changes have been addressed.